File: backend/aiServer/ai_player.py
import math
import random

class PingPongAI:
    def __init__(self, difficulty="medium", custom_settings=None):
        self.games_played = 0
        self.wins = 0
        self.hits = 0
        self.misses = 0

        # Zorluk sevixyesi
        self.difficulty = difficulty
        self.custom_settings = custom_settings
        self.setup_difficulty()

        # Durum takibi
        self.is_frozen = False
        self.target_locked = False
        self.locked_target = None

        # Stratejik kaybetme
        self.consecutive_wins = 0
        self.should_lose_next = False

        # Yeni özellikler
        self.rage_mode = False
        self.rage_counter = 0
        self.tired_mode = False
        self.tired_counter = 0
        self.super_focus = False
        self.focus_counter = 0

		# Yeni -  Oyun alanı bilgileri
        self.screen_width = 800
        self.screen_height = 600
        self.paddle_x = 780  # AI paddle X konumu

    def setup_difficulty(self):
        """Zorluk seviyesine göre parametreleri ayarla"""
        if self.difficulty == "custom" and self.custom_settings:
            # Custom ayarları 10 üzerinden sisteme çevir
            self.reaction_speed = self.custom_settings['reaction_speed'] / 10.0
            self.prediction_accuracy = self.custom_settings['prediction_accuracy'] / 10.0
            self.prepare_distance = 200 + (self.custom_settings['prepare_distance'] * 40)  # 200-600
            self.freeze_distance = 50 + (self.custom_settings['freeze_distance'] * 15)     # 50-200
            self.error_rate = 0.5 - (self.custom_settings['accuracy'] * 0.049)  # 0.5-0.01
            self.learning_rate = self.custom_settings['learning_rate'] / 200.0  # 0.005-0.05
            self.target_win_rate = self.custom_settings['target_win_rate'] / 10.0
            self.lose_probability = (10 - self.custom_settings['fairness']) / 20.0  # 0.5-0.0
            self.max_consecutive_wins = max(1, self.custom_settings['max_consecutive_wins'])

            # özellikler
            self.rage_enabled = self.custom_settings['rage_mode']
            self.fatigue_enabled = self.custom_settings['fatigue_system']
            self.focus_enabled = self.custom_settings['focus_mode']
            self.adaptive_enabled = self.custom_settings['adaptive_difficulty']
            self.prediction_lines = self.custom_settings['show_prediction']

        else:
            # Varsayılan zorluk seviyeleri
            if self.difficulty == "easy":
                self.reaction_speed = 0.5
                self.prediction_accuracy = 0.4
                self.prepare_distance = 300
                self.freeze_distance = 60
                self.error_rate = 0.3
                self.learning_rate = 0.005
                self.target_win_rate = 0.3
                self.lose_probability = 0.4
                self.max_consecutive_wins = 2
            elif self.difficulty == "medium":
                self.reaction_speed = 0.7
                self.prediction_accuracy = 0.7
                self.prepare_distance = 400
                self.freeze_distance = 100
                self.error_rate = 0.15
                self.learning_rate = 0.01
                self.target_win_rate = 0.5
                self.lose_probability = 0.2
                self.max_consecutive_wins = 3
            elif self.difficulty == "hard":
                self.reaction_speed = 0.95
                self.prediction_accuracy = 0.9
                self.prepare_distance = 500
                self.freeze_distance = 140
                self.error_rate = 0.05
                self.learning_rate = 0.02
                self.target_win_rate = 0.7
                self.lose_probability = 0.1
                self.max_consecutive_wins = 5
            elif self.difficulty == "impossible":
                self.reaction_speed = 1.0
                self.prediction_accuracy = 1.0
                self.prepare_distance = 600
                self.freeze_distance = 100
                self.error_rate = 0.00
                self.learning_rate = 0.00
                self.target_win_rate = 1.0
                self.lose_probability = 0.00
                self.max_consecutive_wins = 100

            # Varsayılan özellikler
            self.rage_enabled = False
            self.fatigue_enabled = False
            self.focus_enabled = False
            self.adaptive_enabled = False
            self.prediction_lines = False

    def update_special_modes(self, scored_against_me=False, scored_for_me=False):
        """Özel modları güncelle"""
        # Rage Mode - AI art arda kaybederse öfkelenir ve daha agresif olur
        if self.rage_enabled:
            if scored_against_me:
                self.rage_counter += 1
                if self.rage_counter >= 2:
                    self.rage_mode = True
            elif scored_for_me:
                self.rage_counter = max(0, self.rage_counter - 1)
                if self.rage_counter == 0:
                    self.rage_mode = False

        # Fatigue System - AI çok oynayınca yorulur
        if self.fatigue_enabled:
            if self.games_played > 0 and self.games_played % 5 == 0:
                self.tired_mode = True
                self.tired_counter = 3
            if self.tired_counter > 0:
                self.tired_counter -= 1
                if self.tired_counter == 0:
                    self.tired_mode = False

        # Focus Mode - AI bazen süper odaklanır
        if self.focus_enabled:
            if random.random() < 0.1:  # %10 ihtimal
                self.super_focus = True
                self.focus_counter = 3
            if self.focus_counter > 0:
                self.focus_counter -= 1
                if self.focus_counter == 0:
                    self.super_focus = False

    def get_current_stats(self):
        """Mevcut performans istatistiklerini döndür"""
        current_accuracy = self.prediction_accuracy
        current_error = self.error_rate

        # Rage mode etkisi - sadece doğruluk artar
        if self.rage_mode:
            current_error *= 0.5
            current_accuracy = min(0.98, current_accuracy * 1.2)

        # Tired mode etkisi - performans düşer
        if self.tired_mode:
            current_error *= 1.5
            current_accuracy *= 0.8

        # Focus mode etkisi - mükemmel doğruluk
        if self.super_focus:
            current_accuracy = min(0.99, current_accuracy * 1.3)
            current_error *= 0.3

        return current_accuracy, current_error

    def should_intentionally_lose(self):
        """AI'ın kasten kaybetmesi gerekip gerekmediğini belirler"""
        if self.games_played < 3:
            return False

        # Rage mode'da asla kasten kaybetmez
        if self.rage_mode:
            return False

        current_win_rate = self.wins / max(self.games_played, 1)

        # Adaptive difficulty - oyuncunun performansına göre ayarla
        if hasattr(self, 'adaptive_enabled') and self.adaptive_enabled:
            # Oyuncu çok kaybediyorsa AI daha kolay olsun
            if current_win_rate > 0.8:
                return random.random() < 0.6

        if current_win_rate > self.target_win_rate + 0.2:
            return random.random() < (self.lose_probability * 2)
        elif current_win_rate > self.target_win_rate:
            return random.random() < self.lose_probability
        elif self.consecutive_wins >= self.max_consecutive_wins:
            return random.random() < 0.8

        return False

    def predict_ball_y(self, ball_x, ball_y, ball_speed_x, ball_speed_y, paddle_x, screen_height):
        """Topun paddle_x konumuna geldiğinde Y pozisyonunu tahmin eder"""
		# Hız çok düşükse mevcut Y konumunu döndür
        if abs(ball_speed_x) < 0:
            return ball_y

        # Topun paddle_x konumuna ulaşması için gereken süre
        time_to_reach = (paddle_x - ball_x) / ball_speed_x
        if time_to_reach <= 0:
            return ball_y

        predicted_y = ball_y + ball_speed_y * time_to_reach

        # Yansımaları hesaba kat
        bounces = 0
        while (predicted_y < 0 or predicted_y > screen_height) and bounces < 10:
            if predicted_y < 0:
                predicted_y = -predicted_y
            elif predicted_y > screen_height:
                predicted_y = 2 * screen_height - predicted_y
            bounces += 1

        return predicted_y

    def get_move(self, ball_x, ball_y, ball_speed_x, ball_speed_y,
                 my_paddle_y, paddle_height, screen_height,
                 scored_for_me=False, scored_against_me=False):
        """AI'ın hareket kararını verir - hedef Y konumunu döndürür"""

        if scored_for_me or scored_against_me:
            if scored_for_me:
                self.wins += 1
                self.hits += 1
                self.consecutive_wins += 1
            else:
                self.misses += 1
                self.consecutive_wins = 0
            self.games_played += 1

            # Özel modları güncelle
            self.update_special_modes(scored_against_me, scored_for_me)

            self.is_frozen = False
            self.target_locked = False
            self.locked_target = None
            self.should_lose_next = False

        # Mevcut performans değerlerini al
        current_accuracy, current_error = self.get_current_stats()

        paddle_center = my_paddle_y + paddle_height / 2
        screen_center = screen_height / 2

        # Hedef Y konumu
        target_y = None

        if ball_speed_x > 0 and ball_x > 400:
            if not self.should_lose_next:
                self.should_lose_next = self.should_intentionally_lose()

        if self.should_lose_next and ball_speed_x > 0:
            if random.random() < 0.7:
                predicted_y = self.predict_ball_y(ball_x, ball_y, ball_speed_x, ball_speed_y, 780, screen_height)
                wrong_target = predicted_y + random.choice([-150, 150])
                target_y = max(paddle_height/2, min(screen_height - paddle_height/2, wrong_target))
            else:
                # Mevcut konumda kal
                target_y = paddle_center
        elif random.random() < current_error:
            # Rastgele bir konum seç
            target_y = random.uniform(paddle_height/2, screen_height - paddle_height/2)
        elif ball_speed_x <= 0 or ball_x < 200:
            self.is_frozen = False
            self.target_locked = False
            self.locked_target = None
            # Merkeze git
            target_y = screen_center
        elif ball_speed_x > 0:
            if ball_x > (800 - self.freeze_distance):
                self.is_frozen = True
                # Mevcut konumda kal
                target_y = paddle_center
            elif ball_x > (800 - self.prepare_distance):
                if not self.target_locked:
                    predicted_y = self.predict_ball_y(ball_x, ball_y, ball_speed_x, ball_speed_y, 780, screen_height)
                    noise_range = 50 * (1 - current_accuracy)
                    noise = random.uniform(-noise_range, noise_range)
                    self.locked_target = predicted_y + noise
                    self.target_locked = True

                target_y = self.locked_target
            else:
                predicted_y = self.predict_ball_y(ball_x, ball_y, ball_speed_x, ball_speed_y, 780, screen_height)
                noise_range = 30 * (1 - current_accuracy)
                noise = random.uniform(-noise_range, noise_range)
                target_y = predicted_y + noise
        else:
            # Varsayılan olarak merkeze git
            target_y = screen_center

        # Hedef Y konumunu sınırlar içinde tut
        target_y = max(paddle_height/2, min(screen_height - paddle_height/2, target_y))

		# DEBUG LOG
        print(f"🤖 AI Hesaplama:")
        print(f"  Top: ({ball_x:.1f}, {ball_y:.1f}) | Hız: ({ball_speed_x:.1f}, {ball_speed_y:.1f})")
        print(f"  Paddle Y: {my_paddle_y:.1f} | Hedef: {target_y:.1f}")
        print(f"  Mod: Frozen={self.is_frozen}, Locked={self.target_locked}, Lose={self.should_lose_next}")

        return target_y - paddle_height/2  # Paddle'ın üst kenarının Y konumunu döndür

    def _move_to_target(self, current_pos, target_pos):
        """Belirli bir hedefe hareket et"""
        diff = target_pos - current_pos

        if self.difficulty == "easy":
            threshold = 25
        elif self.difficulty == "medium":
            threshold = 15
        elif self.difficulty == "hard":
            threshold = 8
        elif self.difficulty == "custom":
            threshold = int(25 * (1 - self.prediction_accuracy))
        else:  # impossible
            threshold = 3

        if abs(diff) < threshold:
            return 0
        elif diff > 0:
            return 1
        else:
            return -1

    def _move_to_center(self, current_pos, center_pos):
        """Merkeze hareket et"""
        diff = center_pos - current_pos
        threshold = 30 if self.difficulty == "easy" else 20

        if abs(diff) < threshold:
            return 0
        elif diff > 0:
            return 1
        else:
            return -1

    def update_performance(self, hit_success):
        """Performansa göre parametreleri güncelle"""
        if hit_success:
            self.hits += 1
            if self.difficulty in ["hard", "impossible", "custom"]:
                self.freeze_distance = min(200, self.freeze_distance + 2)
                self.prediction_accuracy = min(0.99, self.prediction_accuracy + self.learning_rate)
        else:
            self.misses += 1
            if self.difficulty in ["hard", "impossible", "custom"]:
                self.freeze_distance = max(80, self.freeze_distance - 3)

    def get_difficulty_info(self):
        """Zorluk seviyesi bilgilerini döndür"""
        difficulty_names = {
            "easy": "KOLAY",
            "medium": "ORTA",
            "hard": "ZOR",
            "impossible": "İMKANSIZ",
            "custom": "ÖZEL"
        }

        return {
            "name": difficulty_names.get(self.difficulty, "ORTA"),
            "reaction": f"{self.reaction_speed:.1f}",
            "accuracy": f"{self.prediction_accuracy:.1f}",
            "error_rate": f"{self.error_rate*100:.0f}%",
            "target_win_rate": f"{self.target_win_rate*100:.0f}%"
        }

    def get_stats(self):
        """İstatistikleri döndür"""
        total_games = max(self.games_played, 1)
        total_attempts = max(self.hits + self.misses, 1)
        current_win_rate = (self.wins / total_games) * 100

        return {
            'games': self.games_played,
            'wins': self.wins,
            'win_rate': current_win_rate,
            'target_win_rate': self.target_win_rate * 100,
            'hit_rate': (self.hits / total_attempts) * 100,
            'hits': self.hits,
            'misses': self.misses,
            'freeze_distance': self.freeze_distance,
            'is_frozen': self.is_frozen,
            'target_locked': self.target_locked,
            'difficulty': self.difficulty,
            'consecutive_wins': self.consecutive_wins,
            'should_lose_next': self.should_lose_next,
            'rage_mode': getattr(self, 'rage_mode', False),
            'tired_mode': getattr(self, 'tired_mode', False),
            'super_focus': getattr(self, 'super_focus', False),
            'rage_counter': getattr(self, 'rage_counter', 0),
            'prediction_lines': getattr(self, 'prediction_lines', False)
        }

    def get_prediction_line(self, ball_x, ball_y, ball_speed_x, ball_speed_y):
        """Tahmin çizgisi için koordinatları döndür"""
        if not getattr(self, 'prediction_lines', False) or ball_speed_x <= 0:
            return None

        predicted_y = self.predict_ball_y(ball_x, ball_y, ball_speed_x, ball_speed_y, 780, 600)
        return predicted_y

File: backend/aiServer/ai_player.js
class PingPongAI {
    constructor(difficulty, customSettings = null) {
        this.gamesPlayed = 0;
        this.wins = 0;
        this.hits = 0;
        this.misses = 0;

        // Zorluk seviyesi
        this.difficulty = difficulty;
        this.customSettings = customSettings;

        this.reactionSpeed = 0.5;
        this.predictionAccuracy = 0.5;
        this.prepareDistance = 300;
        this.freezeDistance = 60;
        this.errorRate = 0.3;
        this.learningRate = 0.01;
        this.targetWinRate = 0.5;
        this.loseProbability = 0.2;
        this.maxConsecutiveWins = 3;
        this.rageEnabled = false;
        this.fatigueEnabled = false;
        this.focusEnabled = false;
        this.adaptiveEnabled = false;
        this.predictionLines = false;

        this.setupDifficulty();

        // Durum takibi
        this.isFrozen = false;
        this.targetLocked = false;
        this.lockedTarget = null;

        // Stratejik kaybetme
        this.consecutiveWins = 0;
        this.shouldLoseNext = false;

        // Yeni özellikler
        this.rageMode = false;
        this.rageCounter = 0;
        this.tiredMode = false;
        this.tiredCounter = 0;
        this.superFocus = false;
        this.focusCounter = 0;

        // Oyun alanı bilgileri
        this.screenWidth = 800;
        this.screenHeight = 600;
        this.paddleX = 780; // AI paddle X konumu
    }

    setupDifficulty() {
        /**Zorluk seviyesine göre parametreleri ayarla*/
        if (this.difficulty === "custom" && this.customSettings) {
            // Frontend'den gelen anahtar adlarını kullan (camelCase)
            let settings = this.customSettings.settings || this.customSettings;

            // Eğer 'settings' key'i varsa onu kullan, yoksa doğrudan customSettings'i kullan
            if (this.customSettings.settings) {
                settings = this.customSettings.settings;
            } else {
                settings = this.customSettings;
            }

            // DOĞRU anahtar adları (camelCase)
            const reactionSpeed = (settings.reactionSpeed !== undefined ? settings.reactionSpeed : 5) / 10.0;
            const predictionAccuracy = (settings.predictionAccuracy !== undefined ? settings.predictionAccuracy : 5) / 10.0;
            const generalAccuracy = (settings.generalAccuracy !== undefined ? settings.generalAccuracy : 5) / 10.0;
            const learningSpeed = (settings.learningSpeed !== undefined ? settings.learningSpeed : 5) / 10.0;
            const prepareDistance = settings.preparationDistance !== undefined ? settings.preparationDistance : 5;
            const freezeDistance = settings.freezeDistance !== undefined ? settings.freezeDistance : 5;
            const targetWinRate = (settings.targetWinRate !== undefined ? settings.targetWinRate : 5) / 10.0;
            const fairness = settings.fairnessLevel !== undefined ? settings.fairnessLevel : 5;
            const maxConsecutive = settings.maxConsecutiveWins !== undefined ? settings.maxConsecutiveWins : 3;

            // Parametreleri ayarla
            this.reactionSpeed = reactionSpeed;
            this.predictionAccuracy = predictionAccuracy;
            this.prepareDistance = 200 + (prepareDistance * 40); // 200-600
            this.freezeDistance = 50 + (freezeDistance * 15);    // 50-200
            this.errorRate = 0.5 - (generalAccuracy * 0.049);    // 0.5-0.01
            this.learningRate = learningSpeed / 200.0;           // 0.005-0.05
            this.targetWinRate = targetWinRate;
            this.loseProbability = (10 - fairness) / 20.0;       // 0.5-0.0
            this.maxConsecutiveWins = Math.max(1, maxConsecutive);

            // Özellikler
            this.rageEnabled = settings.rageMode !== undefined ? settings.rageMode : false;
            this.fatigueEnabled = settings.fatigueSystem !== undefined ? settings.fatigueSystem : false;
            this.focusEnabled = settings.focusMode !== undefined ? settings.focusMode : false;
            this.adaptiveEnabled = settings.adaptiveDifficulty !== undefined ? settings.adaptiveDifficulty : false;
            this.predictionLines = settings.show_prediction !== undefined ? settings.show_prediction : false;

        } else {
            // Varsayılan zorluk seviyeleri
            if (this.difficulty === "easy") {
                this.reactionSpeed = 0.5;
                this.predictionAccuracy = 0.4;
                this.prepareDistance = 300;
                this.freezeDistance = 60;
                this.errorRate = 0.3;
                this.learningRate = 0.005;
                this.targetWinRate = 0.3;
                this.loseProbability = 0.4;
                this.maxConsecutiveWins = 2;
            } else if (this.difficulty === "medium") {
                this.reactionSpeed = 0.7;
                this.predictionAccuracy = 0.7;
                this.prepareDistance = 400;
                this.freezeDistance = 100;
                this.errorRate = 0.15;
                this.learningRate = 0.01;
                this.targetWinRate = 0.5;
                this.loseProbability = 0.2;
                this.maxConsecutiveWins = 3;
            } else if (this.difficulty === "hard") {
                this.reactionSpeed = 0.95;
                this.predictionAccuracy = 0.9;
                this.prepareDistance = 500;
                this.freezeDistance = 140;
                this.errorRate = 0.05;
                this.learningRate = 0.02;
                this.targetWinRate = 0.7;
                this.loseProbability = 0.1;
                this.maxConsecutiveWins = 5;
            } else if (this.difficulty === "impossible") {
                this.reactionSpeed = 1.0;
                this.predictionAccuracy = 1.0;
                this.prepareDistance = 600;
                this.freezeDistance = 100;
                this.errorRate = 0.0;
                this.learningRate = 0.0;
                this.targetWinRate = 1.0;
                this.loseProbability = 0.0;
                this.maxConsecutiveWins = 100;
            }

            // Varsayılan özellikler
            this.rageEnabled = false;
            this.fatigueEnabled = false;
            this.focusEnabled = false;
            this.adaptiveEnabled = false;
            this.predictionLines = false;
        }
    }

    updateSpecialModes(scoredAgainstMe = false, scoredForMe = false) {
        /**Özel modları güncelle*/
        // Rage Mode - AI art arda kaybederse öfkelenir ve daha agresif olur
        if (this.rageEnabled) {
            if (scoredAgainstMe) {
                this.rageCounter += 1;
                if (this.rageCounter >= 2) {
                    this.rageMode = true;
                }
            } else if (scoredForMe) {
                this.rageCounter = Math.max(0, this.rageCounter - 1);
                if (this.rageCounter === 0) {
                    this.rageMode = false;
                }
            }
        }

        // Fatigue System - AI çok oynayınca yorulur
        if (this.fatigueEnabled) {
            if (this.gamesPlayed > 0 && this.gamesPlayed % 5 === 0) {
                this.tiredMode = true;
                this.tiredCounter = 3;
            }
            if (this.tiredCounter > 0) {
                this.tiredCounter -= 1;
                if (this.tiredCounter === 0) {
                    this.tiredMode = false;
                }
            }
        }

        // Focus Mode - AI bazen süper odaklanır
        if (this.focusEnabled) {
            if (Math.random() < 0.1) { // %10 ihtimal
                this.superFocus = true;
                this.focusCounter = 3;
            }
            if (this.focusCounter > 0) {
                this.focusCounter -= 1;
                if (this.focusCounter === 0) {
                    this.superFocus = false;
                }
            }
        }
    }

    getCurrentStats() {
        /**Mevcut performans istatistiklerini döndür*/
        let currentAccuracy = this.predictionAccuracy;
        let currentError = this.errorRate;

        // Rage mode etkisi - sadece doğruluk artar
        if (this.rageMode) {
            currentError *= 0.5;
            currentAccuracy = Math.min(0.98, currentAccuracy * 1.2);
        }

        // Tired mode etkisi - performans düşer
        if (this.tiredMode) {
            currentError *= 1.5;
            currentAccuracy *= 0.8;
        }

        // Focus mode etkisi - mükemmel doğruluk
        if (this.superFocus) {
            currentAccuracy = Math.min(0.99, currentAccuracy * 1.3);
            currentError *= 0.3;
        }

        return [currentAccuracy, currentError];
    }

    shouldIntentionallyLose() {
        /**AI'ın kasten kaybetmesi gerekip gerekmediğini belirler*/
        if (this.gamesPlayed < 3) {
            return false;
        }

        // Rage mode'da asla kasten kaybetmez
        if (this.rageMode) {
            return false;
        }

        const currentWinRate = this.wins / Math.max(this.gamesPlayed, 1);

        // Adaptive difficulty - oyuncunun performansına göre ayarla
        if (this.adaptiveEnabled) {
            // Oyuncu çok kaybediyorsa AI daha kolay olsun
            if (currentWinRate > 0.8) {
                return Math.random() < 0.6;
            }
        }

        if (currentWinRate > this.targetWinRate + 0.2) {
            return Math.random() < (this.loseProbability * 2);
        } else if (currentWinRate > this.targetWinRate) {
            return Math.random() < this.loseProbability;
        } else if (this.consecutiveWins >= this.maxConsecutiveWins) {
            return Math.random() < 0.8;
        }

        return false;
    }

    predictBallY(ballX, ballY, ballSpeedX, ballSpeedY, paddleX, screenHeight) {
        /**Topun paddle_x konumuna geldiğinde Y pozisyonunu tahmin eder*/
        // Hız çok düşükse mevcut Y konumunu döndür
        if (Math.abs(ballSpeedX) < 0.1) {
            return ballY;
        }

        // Topun paddle_x konumuna ulaşması için gereken süre
        const timeToReach = (paddleX - ballX) / ballSpeedX;
        if (timeToReach <= 0) {
            return ballY;
        }

        let predictedY = ballY + ballSpeedY * timeToReach;

        // Yansımaları hesaba kat
        let bounces = 0;
        while ((predictedY < 0 || predictedY > screenHeight) && bounces < 10) {
            if (predictedY < 0) {
                predictedY = -predictedY;
            } else if (predictedY > screenHeight) {
                predictedY = 2 * screenHeight - predictedY;
            }
            bounces += 1;
        }

        return predictedY;
    }

    getMove(ballX, ballY, ballSpeedX, ballSpeedY, myPaddleY, paddleHeight, screenHeight, scoredForMe = false, scoredAgainstMe = false) {
        /**AI'ın hareket kararını verir - hedef Y konumunu döndürür*/

        if (scoredForMe || scoredAgainstMe) {
            if (scoredForMe) {
                this.wins += 1;
                this.hits += 1;
                this.consecutiveWins += 1;
            } else {
                this.misses += 1;
                this.consecutiveWins = 0;
            }
            this.gamesPlayed += 1;

            // Özel modları güncelle
            this.updateSpecialModes(scoredAgainstMe, scoredForMe);

            this.isFrozen = false;
            this.targetLocked = false;
            this.lockedTarget = null;
            this.shouldLoseNext = false;
        }

        // Mevcut performans değerlerini al
        const [currentAccuracy, currentError] = this.getCurrentStats();

        const paddleCenter = myPaddleY + paddleHeight / 2;
        const screenCenter = screenHeight / 2;

        // Hedef Y konumu
        let targetY = null;

        if (ballSpeedX > 0 && ballX > 400) {
            if (!this.shouldLoseNext) {
                this.shouldLoseNext = this.shouldIntentionallyLose();
            }
        }

        if (this.shouldLoseNext && ballSpeedX > 0) {
            if (Math.random() < 0.7) {
                const predictedY = this.predictBallY(ballX, ballY, ballSpeedX, ballSpeedY, 780, screenHeight);
                const wrongTarget = predictedY + (Math.random() < 0.5 ? -150 : 150);
                targetY = Math.max(paddleHeight / 2, Math.min(screenHeight - paddleHeight / 2, wrongTarget));
            } else {
                // Mevcut konumda kal
                targetY = paddleCenter;
            }
        } else if (Math.random() < currentError) {
            // Rastgele bir konum seç
            targetY = Math.random() * (screenHeight - paddleHeight) + paddleHeight / 2;
        } else if (ballSpeedX <= 0 || ballX < 200) {
            this.isFrozen = false;
            this.targetLocked = false;
            this.lockedTarget = null;
            // Merkeze git
            targetY = screenCenter;
        } else if (ballSpeedX > 0) {
            if (ballX > (800 - this.freezeDistance)) {
                this.isFrozen = true;
                // Mevcut konumda kal
                targetY = paddleCenter;
            } else if (ballX > (800 - this.prepareDistance)) {
                if (!this.targetLocked) {
                    const predictedY = this.predictBallY(ballX, ballY, ballSpeedX, ballSpeedY, 780, screenHeight);
                    const noiseRange = 50 * (1 - currentAccuracy);
                    const noise = Math.random() * (noiseRange * 2) - noiseRange;
                    this.lockedTarget = predictedY + noise;
                    this.targetLocked = true;
                }

                targetY = this.lockedTarget;
            } else {
                const predictedY = this.predictBallY(ballX, ballY, ballSpeedX, ballSpeedY, 780, screenHeight);
                const noiseRange = 30 * (1 - currentAccuracy);
                const noise = Math.random() * (noiseRange * 2) - noiseRange;
                targetY = predictedY + noise;
            }
        } else {
            // Varsayılan olarak merkeze git
            targetY = screenCenter;
        }

        // Hedef Y konumunu sınırlar içinde tut
        targetY = Math.max(paddleHeight / 2, Math.min(screenHeight - paddleHeight / 2, targetY));

        // DEBUG LOG
        console.log(`🤖 AI Hesaplama:`);
        console.log(`  Top: (${ballX.toFixed(1)}, ${ballY.toFixed(1)}) | Hız: (${ballSpeedX.toFixed(1)}, ${ballSpeedY.toFixed(1)})`);
        console.log(`  Paddle Y: ${myPaddleY.toFixed(1)} | Hedef: ${targetY.toFixed(1)}`);
        console.log(`  Mod: Frozen=${this.isFrozen}, Locked=${this.targetLocked}, Lose=${this.shouldLoseNext}`);

        return targetY - paddleHeight / 2; // Paddle'ın üst kenarının Y konumunu döndür
    }

    _moveToTarget(currentPos, targetPos) {
        /**Belirli bir hedefe hareket et*/
        const diff = targetPos - currentPos;

        let threshold;
        if (this.difficulty === "easy") {
            threshold = 25;
        } else if (this.difficulty === "medium") {
            threshold = 15;
        } else if (this.difficulty === "hard") {
            threshold = 8;
        } else if (this.difficulty === "custom") {
            threshold = Math.floor(25 * (1 - this.predictionAccuracy));
        } else { // impossible
            threshold = 3;
        }

        if (Math.abs(diff) < threshold) {
            return 0;
        } else if (diff > 0) {
            return 1;
        } else {
            return -1;
        }
    }

    _moveToCenter(currentPos, centerPos) {
        /**Merkeze hareket et*/
        const diff = centerPos - currentPos;
        const threshold = this.difficulty === "easy" ? 30 : 20;

        if (Math.abs(diff) < threshold) {
            return 0;
        } else if (diff > 0) {
            return 1;
        } else {
            return -1;
        }
    }

    updatePerformance(hitSuccess) {
        /**Performansa göre parametreleri güncelle*/
        if (hitSuccess) {
            this.hits += 1;
            if (["hard", "impossible", "custom"].includes(this.difficulty)) {
                this.freezeDistance = Math.min(200, this.freezeDistance + 2);
                this.predictionAccuracy = Math.min(0.99, this.predictionAccuracy + this.learningRate);
            }
        } else {
            this.misses += 1;
            if (["hard", "impossible", "custom"].includes(this.difficulty)) {
                this.freezeDistance = Math.max(80, this.freezeDistance - 3);
            }
        }
    }

    getDifficultyInfo() {
        /**Zorluk seviyesi bilgilerini döndür*/
        const difficultyNames = {
            "easy": "KOLAY",
            "medium": "ORTA",
            "hard": "ZOR",
            "impossible": "İMKANSIZ",
            "custom": "ÖZEL"
        };

        return {
            name: difficultyNames[this.difficulty] || "ORTA",
            reaction: this.reactionSpeed.toFixed(1),
            accuracy: this.predictionAccuracy.toFixed(1),
            error_rate: `${(this.errorRate * 100).toFixed(0)}%`,
            target_win_rate: `${(this.targetWinRate * 100).toFixed(0)}%`
        };
    }

    getStats() {
        /**İstatistikleri döndür*/
        const totalGames = Math.max(this.gamesPlayed, 1);
        const totalAttempts = Math.max(this.hits + this.misses, 1);
        const currentWinRate = (this.wins / totalGames) * 100;

        return {
            games: this.gamesPlayed,
            wins: this.wins,
            win_rate: currentWinRate,
            target_win_rate: this.targetWinRate * 100,
            hit_rate: (this.hits / totalAttempts) * 100,
            hits: this.hits,
            misses: this.misses,
            freeze_distance: this.freezeDistance,
            is_frozen: this.isFrozen,
            target_locked: this.targetLocked,
            difficulty: this.difficulty,
            consecutive_wins: this.consecutiveWins,
            should_lose_next: this.shouldLoseNext,
            rage_mode: this.rageMode,
            tired_mode: this.tiredMode,
            super_focus: this.superFocus,
            rage_counter: this.rageCounter,
            prediction_lines: this.predictionLines
        };
    }

    getPredictionLine(ballX, ballY, ballSpeedX, ballSpeedY) {
        /**Tahmin çizgisi için koordinatları döndür*/
        if (!this.predictionLines || ballSpeedX <= 0) {
            return null;
        }

        const predictedY = this.predictBallY(ballX, ballY, ballSpeedX, ballSpeedY, 780, 600);
        return predictedY;
    }
}

// Node.js için export (eğer gerekirse)
if (typeof module !== 'undefined' && module.exports) {
    module.exports = PingPongAI;
}
